### ISSUE TITLE
Record/Node Reordering in DataGrid and TreeList and Cross-Component Drag and Drop

### ISSUE THREAD

**Issue by Alyar666** [DevExpress/DevExtreme#9339]
# The Problem

 The **DataGrid** and **TreeList** do not allow users to drag and drop rows. Drag and drop between components is also unsupported.

 # The Proposed Solution

 We plan to introduce the **Draggable** and **Sortable** components. Refer to [New Drag and Drop Utilities](https://github.com/DevExpress/DevExtreme/issues/9338) for more information.

 We will integrate these components into the **DataGrid**, **TreeList**, **List**, and **Scheduler** to enable cross-component drag and drop operations and also allow rows to be dragged and dropped in the **DataGrid** and **TreeList**.

![DataGrid_Drag_and_Drop_rows](https://user-images.githubusercontent.com/23191430/63279161-34106380-c2b1-11e9-9763-be04538f3723.gif)

 The following API is planned:

 - **DataGrid** and **TreeList**

     ```js
    let options = {
        rowDragging: {
            allowReordering: boolean,
            allowDropInsideItem: boolean,
            showDragIcons: boolean,
            group: string,
            data: any,
            onDragStart: function,
            onDragMove: function,
            onDragChange: function,
            onDragEnd: function,
            onReorder: function,
            onAdd: function,
            onRemove: function
        }
    };
    ```
 - **Scheduler**
     ```js
    let schedulerOptions = {
        appointmentDragging: {
            group: string,
            data: string,
            onDragStart: function,
            onDragMove: function,
            onDragEnd: function,
            onAdd: function,
            onRemove: function
        }
    };
    ```
 - **List**
     ```js
    let listOptions = {
        itemDragging: {
            allowReordering: boolean,
            group: string,
            data: any,
            onDragStart: function,
            onDragMove: function,
            onDragChange: function,
            onDragEnd: function,
            onReorder: function,
            onAdd: function,
            onRemove: function
        }
    };
    ```
 ## Drag and drop rows between two grids
 ```js
$("#grid1").dxDataGrid({
    rowDragging: {
        allowReordering: true,
        group: "shared",
        data: "grid1"
    }
});
 $("#grid2").dxDataGrid({
    rowDragging: {
        allowReordering: true,
        group: "shared",
        data: "grid2"
    }
});
```
 ## Drag and drop elements between a Scheduler and a List
 ```js
$("#scheduler").dxScheduler({
    appointmentDragging: {
        group: "shared",
        data: "scheduler"
    }
});
 $("#list").dxList({
    itemDragging: {
        group: "shared",
        data: "list"
    }
});
```
 # We Need Your Feedback
 ## Take a Quick Poll
 [Do you find record/node reordering and cross-component drag and drop useful?](https://docs.google.com/forms/d/e/1FAIpQLSdvCnNZeF-466I4OEcgPYXoftxIX9sppxkCrlaMGtTSVGO5Kg/viewform)
 ## Live Sandbox
DataGrid row reordering: [JQuery](https://codepen.io/vconst/pen/dyyyqXJ), [Angular](https://codesandbox.io/s/local-reordering-devextreme-data-grid-jjbes), [Vue](https://codesandbox.io/s/local-reordering-devextreme-data-grid-h0w68), [React](https://codesandbox.io/s/local-reordering-devextreme-data-grid-gx3t3)
Drag and drop rows between DataGrids: [JQuery](https://codepen.io/vconst/pen/QWWWJrJ)
TreeList multiple node dragging: [JQuery](https://codepen.io/vconst/pen/wvvvYGM), [Angular](https://codesandbox.io/s/tree-list-node-dragging-lw7fu)
TreeView node dragging: [JQuery](https://codepen.io/vconst/pen/mdddQVR), [Angular](https://codesandbox.io/s/plain-data-structure-devextreme-tree-view-6qf5j), [React](https://codesandbox.io/s/plain-data-structure-devextreme-tree-view-ecdb7)
 # Get Notified of Updates
 Subscribe to this thread or to our [Facebook](https://www.facebook.com/DevExpress.DevExtreme/) and [Twitter](https://twitter.com/devextreme) accounts for updates on this topic.

**Comment by Bykiev** [comment]
@Alyar666, hi!

Seems very promising, but we need a live demo to touch it :З Drag-and-drop feature is must have, also we need reordering feature for a treeView component. To achive this right now we're using our custom solution with jQuery sortable.

**Comment by Alyar666** [comment]
Hello @Bykiev,

> Seems very promising, but we need a live demo to touch it 

At the moment, we focus our forces on its implementation. We will do our best to show the basic functionality as soon as possible. I'll update this issue when the testing samples are ready.
 
> Drag-and-drop feature is must have, also we need reordering feature for a treeView component.

According to the current state of development, I am afraid that we may not have time to release it in v19.2 with the desired level of quality. It's at the top of our to-do list, so if it's not ready by v19.2, we'll release it in future versions. Please bear with us.

**Comment by mcardoalm** [comment]
We really need drag and drop for TreeView as well, if it can make it into 19.2 it would be a huge help!!!

**Comment by chaddblanchard** [comment]
+1 for drag and drop in TreeView. This would be a great addition to 19.2!

**Comment by krilllind** [comment]
Ok so this is great news, but in practice you also have to think about what applications this will be used in.

Let's say for example that you want to enable DnD, will you also allow for multiselect in combination? And if so, what happens when you enable virtual scrolling for TreeList (due to performance issues). Will you then be able to scroll down, select some nodes and then DnD these selected nodes up/down the TreeList?

**Comment by Bykiev** [comment]
@Alyar666, will it be possible to allow reordering only same-level items in treeList (in jQuery sortable it can be achieved by using `containment` property)?

**Comment by vconst** [comment]
We have published 19.2.2-beta packages. I have added some live examples to the main content. You're welcome to try them and share your feedback here.

@mcardoalm @blanchardc While TreeView doesn't support D&D out of the box, it's possible to use our new dxSortable component to manually complete this task. I have created an example demonstrating this idea - [JQuery](https://codepen.io/vconst/pen/mdddQVR), [Angular](https://codesandbox.io/s/plain-data-structure-devextreme-tree-view-6qf5j), [React](https://codesandbox.io/s/plain-data-structure-devextreme-tree-view-ecdb7)

@krilllind, This feature supports Virtual Scrolling out of the box. Also, you can implement multiple nodes D&D by handing onReorder. See the following examples - [JQuery](https://codepen.io/vconst/pen/wvvvYGM), [Angular](https://codesandbox.io/s/tree-list-node-dragging-lw7fu)

@Bykiev, You can allow reordering of only same-level items by handing onDragChange. See the following example - [https://codepen.io/vconst/pen/xxxGpKG](https://codepen.io/vconst/pen/xxxGpKG)

**Comment by krilllind** [comment]
@vconst The DnD example is working fine, will we be able to combine this with reordering, and not just within the same parent row, but drag outside parent and drop between some other rows?

Some feedback on reordering!

It's currently hard to see where a row will be inserted after reorder, would be even better if you could enhance that feedback to the user draging the row. Maybe apply some opacity to the row you are draging?

Btw: Nice to see that multiselct is supported ⭐️

**Comment by vconst** [comment]
@krilllind, The rowDragging.allowReordering option can be used for TreeList. See the following example: [https://codepen.io/vconst/pen/gOOaQmp](https://codepen.io/vconst/pen/gOOaQmp)

Also, we are going to add the opacity 0.9 to dragging row in the release version. I have added a style with this opacity to the previous example.

**Comment by markeaston63** [comment]
I tried it but keep getting TypeError: Cannot set property 'dxSortable' of undefined. I have nothing in grid configuration that I would see would cause that. ie 

`<dx-data-grid #datagrid class="ml-1 mt-1 mr-1 mb-1"
    [dataSource]="assemblyList"
    keyExpr="id"
    noDataText="Click the Generate button to generate the assemblies (from the BOM)"
    [showBorders]="true"
    [showRowLines]="true"
    [height]="600"
    (onEditingStart)="onEditingStart($event)"
    (onRowUpdated)="onRowUpdated($event)"
    (onRowClick)="onRowClick($event)"
    >
    <dxo-row-dragging
        [allowReordering]="true"
        [showDragIcons]="true"
        [onDragStart]="onDragStart"
        [onReorder]="onReorder"
    ></dxo-row-dragging>
    <dxo-scrolling mode="virtual"></dxo-scrolling>
    <dxo-sorting mode="none"></dxo-sorting>
    <dxo-editing mode="batch" [allowUpdating]="true" [allowDeleting]="false" [useIcons]="true"></dxo-editing>
    <dxi-column name="No" cellTemplate="NoTemplate" width="40px" fixed="true" caption="Row"></dxi-column>
    <dxi-column dataField="number" [allowEditing]="false"  width="80px" fixed="true" caption="Number"></dxi-column>
    <dxi-column dataField="type"  [allowEditing]="false" width="60px" caption="Type"></dxi-column>
    <dxi-column dataField="sortSequence" [allowEditing]="true" width="110px" caption="Your Sequence"></dxi-column>
    <dxi-column dataField="assemblyTypeSequence" [allowEditing]="false" width="110px" caption="Type Sequence"></dxi-column>
    <dxi-column dataField="stackSequence"  [allowEditing]="true" width="120px" caption="Stack Sequence"></dxi-column>
    <dxi-column dataField="zone"  [allowEditing]="true" [width]="60" caption="Zone"></dxi-column>
    <dxi-column dataField="level"  [allowEditing]="false" [width]="60" caption="Level"></dxi-column>
    <dxi-column dataField="trussSpan"  [allowEditing]="false" [width]="80" caption="Truss Span"></dxi-column>
    <dxi-column dataField="thumbnail" [width]="70" [allowEditing]="false" cellTemplate="thumbnailTemplate" caption="Sheet"></dxi-column>
    <div *dxTemplate="let data of 'NoTemplate'">
        {{ data.component.pageIndex() * data.component.pageSize() + data.rowIndex + 1 }}
    </div>
    <div *dxTemplate="let data of 'thumbnailTemplate'">
        <img [src]="'data:image/jpeg;base64,' + data.value" style="height: 100px;display: block;"/>
    </div>
</dx-data-grid>`

**Comment by markeaston63** [comment]
Hi, I found a bug - but works great otherwise ...

This does not work:

<dx-data-grid
#datagrid
id="gridContainer"
[height]="600"
[dataSource]="assemblyList"
[showBorders]="true">
<dxo-row-dragging
[allowReordering]="true"
[showDragIcons]="true"
[onDragStart]="onDragStart"
[onReorder]="onReorder"
>


<dxo-editing mode="batch" [allowUpdating]="true" [allowDeleting]="false" [useIcons]="true">
<dxi-column dataField="assemblyNumber" [allowEditing]="false" width="80px" fixed="true" caption="Assembly No">
<dxi-column dataField="assemblyType" [allowEditing]="false" caption="Aseembly Type">
And this does work:

<dx-data-grid
#datagrid
id="gridContainer"
[height]="600"
[dataSource]="assemblyList"
[showBorders]="true">
<dxo-row-dragging
[allowReordering]="true"
[showDragIcons]="true"
[onDragStart]="onDragStart"
[onReorder]="onReorder"
>


<dxo-editing mode="batch" [allowUpdating]="true" [allowDeleting]="false" [useIcons]="true">
<dxi-column dataField="assemblyNumber" [allowEditing]="false" width="80px" caption="Assembly No">
<dxi-column dataField="assemblyType" [allowEditing]="false" caption="Aseembly Type">
It fails on fixed="true"​

**Comment by sometimesElen** [comment]
Hello,

We are aware of this issue and will research it in the context of the [DataGrid with Drag and Drop raises "TypeError: Cannot set property 'dxSortable' of undefined" when at least one column is fixed](https://www.devexpress.com/Support/Center/Question/Details/T824556/datagrid-with-drag-and-drop-raises-typeerror-cannot-set-property-dxsortable-of-undefined) ticket. 

Please bear with us.

**Comment by vconst** [comment]
@markeaston63, We have already fixed the issue in 19.2.3.

**Comment by JeanMeche** [comment]
I just found another bug : Enabling row dragging & Master/Detail at the same time. 

When a detail is expanded, dragging the following row, it's the detail that gets dragged !

**Comment by Alyar666** [comment]
Hello @kyro38,

Thank you for contacting us. We will process this issue in the context of the following Support Center thread:
[DataGrid - The row dragging doesn't work correctly when there is master-detail](https://www.devexpress.com/Support/Center/Question/Details/T828366)

**Comment by Alyar666** [comment]
These features are now available in the [v19.2](https://github.com/DevExpress/DevExtreme/releases/tag/19.2.3) release. I'm closing this thread. In the case of bugs or questions, feel free to create new GitHub [issues](https://github.com/DevExpress/DevExtreme/issues) or tickets in our [Support Center](https://www.devexpress.com/Support/Center/).
